qutip/scattering.py

```diff
@@ -0,0 +1,398 @@
+
+        Returns
+        -------
+        count: int
+            the number of propagators which have been computed so far
+        """
+        count = 0
+        for _, dic in self.propagators.items():
+            for _, prop in dic.items():
+                if prop is not None:
+                    count += 1
+        return count
+
+
+def set_partition(collection, num_sets):
+    """
+    Enumerate all ways of partitioning collection into num_sets different lists,
+    e.g. list(set_partition([1,2], 2)) = [[[1, 2], []], [[1], [2]], [[2], [1]],
+    [[], [1, 2]]]
+
+    Parameters
+    ----------
+    collection : iterable
+        collection to generate a set partition of
+    num_sets : int
+        number of sets to partition collection into
+
+    Returns
+    -------
+    partition : iterable
+        the partitioning of collection into num_sets sets
+    """
+    for partitioning in product(range(num_sets), repeat = len(collection)):
+        partition = [[] for _ in range(num_sets)]
+        for i, set_index in enumerate(partitioning):
+            partition[set_index].append(collection[i])
+        yield tuple(tuple(indices) for indices in partition)
+
+
+def photon_scattering_operator(evolver, c_ops, taus_list):
+    """
+    Compute the scattering operator for a system emitting into multiple
+    waveguides.
+
+    Parameters
+    ----------
+    evolver : Evolver
+        Evolver-wrapped Hamiltonian describing the system
+    c_ops : list
+        list of collapse operators for each waveguide; these are assumed to
+        include spontaneous decay rates, e.g.
+        :math:`\\sigma = \\sqrt \\gamma \\cdot a`
+    taus_list : list-like
+        list of (list of emission times) for each waveguide
+
+    Returns
+    -------
+    Omega : Qobj
+        The temporal scattering operator with dimensionality equal to the
+        system state
+    """
+
+    omega = 1
+
+    # Extract the full list of taus
+    taus = [(0.0, -1)]  # temporal "ground state" for arbitrary waveguide
+    for i, tauWG in enumerate(taus_list):
+        for tau in tauWG:
+            taus.append((tau, i))
+    taus.sort(key = lambda tup: tup[0])  # sort taus by time
+
+    # Compute Prod Ueff(tq, tq-1)
+    for i in range(1, len(taus)):
+        tq, q = taus[i]
+        tprev, _ = taus[i - 1]
+        omega = c_ops[q] * evolver.prop(tq, tprev) * omega
+
+    # Add the <0|Uff(TP, tm)|0> term
+    tmax = evolver.tlist[-1]
+    taumax, _ = taus[-1]
+    # if taus[-1] < tmax:
+    omega = evolver.prop(tmax, taumax) * omega
+
+    return omega
+
+
+def temporal_basis_vector(waveguide_emission_indices, n_time_bins):
+    """
+    Generate a temporal basis vector for emissions at specified time bins into
+    specified waveguides
+
+    Parameters
+    ----------
+    waveguide_emission_indices : list or tuple
+        list of indices where photon emission occurs for each waveguide,
+        e.g. [[t_1]_wg1, [t_1, t_2]_wg2, []_wg3, [t_1, t_2, t_3]_wg4]
```

Comment:
Will make the proposed change to the docstring. Making this a dictionary rather than a list of lists would involve significant effort and less clearly demonstrates the relationship between the emission time partitionings and the way that the temporal basis is constructed (see "The temporal basis" section of the notebook in my [related PR](https://github.com/qutip/qutip-notebooks/pull/58))